src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java

```diff
@@ -636,51 +672,93 @@ private HttpEntity newStashBuildNotificationEntity(
         json.put("description", getBuildDescription(build, state));
         json.put("url", Jenkins.getInstance()
         		.getRootUrl().concat(build.getUrl()));
-        
+
         return new StringEntity(json.toString(), "UTF-8");
 	}
 
 	/**
-	 * Returns the build key used in the Stash notification. Includes the 
-	 * build number depending on the user setting.
-	 * 
-	 * @param 	build	the build to notify Stash of
-	 * @return	the build key for the Stash notification
+	 * Return the old-fashion build key
+	 *
+	 * @param  build the build to notify Stash of
+	 * @return default build key
 	 */
-	private String getBuildKey(final AbstractBuild<?, ?> build) {
+	private String getDefaultBuildKey(final AbstractBuild<?, ?> build) {
 		StringBuilder key = new StringBuilder();
+
 		key.append(build.getProject().getName());
-        if (includeBuildNumberInKey 
-        		|| getDescriptor().isIncludeBuildNumberInKey()) {
+		if (includeBuildNumberInKey
+				|| getDescriptor().isIncludeBuildNumberInKey()) {
 			key.append('-').append(build.getNumber());
 		}
 		key.append('-').append(Jenkins.getInstance().getRootUrl());
+
+		return key.toString();
+	}
+
+		/**
+         * Returns the build key used in the Stash notification. Includes the
+         * build number depending on the user setting.
+         *
+         * @param 	build	the build to notify Stash of
+         * @return	the build key for the Stash notification
+         */
+	private String getBuildKey(final AbstractBuild<?, ?> build,
+							   BuildListener listener) {
+
+		StringBuilder key = new StringBuilder();
+
+		if (prependParentProjectKey || getDescriptor().isPrependParentProjectKey()){
+			if (null != build.getParent().getParent()) {
+				key.append(build.getParent().getParent().getFullName()).append('-');
+			}
+		}
+
+		String overriddenKey = (projectKey != null && projectKey.trim().length() > 0) ? projectKey : getDescriptor().getProjectKey();
```

Comment:
I'm not sure it is the best way to check that parameter is not empty. Is this correct? projectKey is never set up at global configuration, so I can miss something due to I'm new to Jenkins API